esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,268 @@
+  const [packageManifestLoc, packageManifest] = resolveDependencyManifest(packageName, cwd);
+
+  if (packageManifestLoc && packageManifest && typeof packageManifest.exports !== 'undefined') {
+    // If this is a non-main entry point
+    if (packageEntrypoint) {
+      const exportMapValue = resolveExportsMap(
+        packageManifest as PackageManifestWithExports,
+        './' + packageEntrypoint,
+      );
+
+      if (typeof exportMapValue !== 'string') {
+        throw new Error(
+          `Package "${packageName}" exists but package.json "exports" does not include entry for "./${packageEntrypoint}".`,
+        );
+      }
+      return path.join(packageManifestLoc, '..', exportMapValue);
+    } else {
+      const exportMapValue = resolveExportsMap(packageManifest as PackageManifestWithExports, '.');
+
+      if (exportMapValue) {
+        return path.join(packageManifestLoc, '..', exportMapValue);
+      }
+    }
+  }
+
+  // if, no export map and dep points directly to a file within a package, return that reference.
+  if (path.extname(dep) && !validatePackageName(dep).validForNewPackages) {
+    return fs.realpathSync.native(require.resolve(dep, {paths: [cwd]}));
+  }
+
+  // Otherwise, resolve directly to the dep specifier. Note that this supports both
+  // "package-name" & "package-name/some/path" where "package-name/some/path/package.json"
+  // exists at that lower path, that must be used to resolve. In that case, export
+  // maps should not be supported.
+  const [depManifestLoc, depManifest] = resolveDependencyManifest(dep, cwd);
+
+  if (!depManifest) {
+    try {
+      const maybeLoc = fs.realpathSync.native(require.resolve(dep, {paths: [cwd]}));
+      return maybeLoc;
+    } catch (err) {
+      // Oh well, was worth a try
+    }
+  }
+
+  if (!depManifestLoc || !depManifest) {
+    throw new Error(
+      `Package "${dep}" not found. Have you installed it? ${depManifestLoc ? depManifestLoc : ''}`,
+    );
+  }
+  if (
+    depManifest.name &&
+    (depManifest.name.startsWith('@reactesm') || depManifest.name.startsWith('@pika/react'))
+  ) {
+    throw new Error(
+      `React workaround packages no longer needed! Revert back to the official React & React-DOM packages.`,
+    );
+  }
+  let foundEntrypoint = resolveManifestEntry(depManifest, dep, {
+    packageName,
+    packageLookupFields,
+  });
+
+  // Some packages are types-only. If this is one of those packages, resolve with that.
+  if (!foundEntrypoint && manifestHasTypes(depManifest)) {
+    const typesLoc = (depManifest.types || depManifest.typings) as string;
+    return path.join(depManifestLoc, '..', typesLoc);
+  }
+  // Sometimes packages don't give an entrypoint, assuming you'll fall back to "index.js".
+  if (!foundEntrypoint) {
+    foundEntrypoint = 'index.js';
+  }
+  if (typeof foundEntrypoint !== 'string') {
+    throw new Error(`"${dep}" has unexpected entrypoint: ${JSON.stringify(foundEntrypoint)}.`);
+  }
+
+  return fs.realpathSync.native(
```

Comment:
`"main": "./foo"` can also resolve to `./foo/index.js` if foo is a directory. Make sure that that is tested and handled if it isn't yet.